### PR TITLE
promql: Limit extrapolation of delta/rate/increase

### LINKED ISSUE #581: `rate` and `delta` deal poorly with slow moving data.
For data that typically changes less often than the sampling window, the delta expansion to avoid aliasing actually add inaccuracy.

For example, there is a spike in the rate to 80 (see http://bit.ly/1A4suIw) which is an order of magnitude greater than the actual rate (which is <10).

### PR DESCRIPTION
**Pull request by beorn7**
Here the promised PR. I'm planning to write a blog post about the whole rationale behind the extrapolation… Will be fun.

The new implementation detects the start and end of a series by
looking at the average sample interval within the range. If the first
(last) sample in the range is more than 1.1*interval distant from the
beginning (end) of the range, it is considered the first (last) sample
of the series as a whole, and extrapolation is limited to half the
interval (rather than all the way to the beginning (end) of the
range). In addition, if the extrapolated starting point of a counter
(where it is zero) is within the range, it is used as the starting
point of the series.

Fixes #581 

@brian-brazil

### CODE REVIEW

**Comment by brian-brazil** [comment]
:+1:

File: promql/functions.go [hunk 1]
```diff
@@ -79,22 +75,48 @@ func funcDelta(ev *evaluator, args Expressions) model.Value {
...
+
+		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp).Seconds()
+		averageDurationBetweenSamples := sampledInterval / float64(len(samples.Values)-1)
+
+		// If the first/last samples are close to the boundaries of the range,
+		// extrapolate the result. This is as we expect that another sample
+		// will exist given the spacing between samples we've seen thus far,
+		// with an allowance for noise.
+		extrapolationThreshold := averageDurationBetweenSamples * 1.1
+		extrapolateToInterval := sampledInterval
```
Comment (grobie) [inline promql/functions.go]: Would it be possible to move these lines+comment below the next if block? I found it hard to follow to read about extrapolation, then negative counter capping, then back to extrapolation.

**Comment by beorn7** [comment]
Merging. \o/

File: promql/functions.go (hunk 1 above)
Comment (beorn7) [inline promql/functions.go]: Sure...
